File: src/daemon/daemon.cpp
// Copyright (c) 2014-2024, The Monero Project
// 
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without modification, are
// permitted provided that the following conditions are met:
//
// 1. Redistributions of source code must retain the above copyright notice, this list of
//    conditions and the following disclaimer.
//
// 2. Redistributions in binary form must reproduce the above copyright notice, this list
//    of conditions and the following disclaimer in the documentation and/or other
//    materials provided with the distribution.
//
// 3. Neither the name of the copyright holder nor the names of its contributors may be
//    used to endorse or promote products derived from this software without specific
//    prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
//
// Parts of this file are originally copyright (c) 2012-2013 The Cryptonote developers

#include <memory>
#include <stdexcept>
#include <boost/algorithm/string/split.hpp>
#include "misc_log_ex.h"
#include "daemon/daemon.h"
#include "rpc/daemon_handler.h"
#include "rpc/zmq_pub.h"
#include "rpc/zmq_server.h"

#include "common/password.h"
#include "common/util.h"
#include "cryptonote_basic/events.h"
#include "daemon/core.h"
#include "daemon/p2p.h"
#include "daemon/protocol.h"
#include "daemon/rpc.h"
#include "daemon/command_server.h"
#include "daemon/command_line_args.h"
#include "net/net_ssl.h"
#include "version.h"

using namespace epee;

#include <functional>

#undef MONERO_DEFAULT_LOG_CATEGORY
#define MONERO_DEFAULT_LOG_CATEGORY "daemon"

namespace daemonize {

struct zmq_internals
{
  explicit zmq_internals(t_core& core, t_p2p& p2p)
    : rpc_handler{core.get(), p2p.get()}
    , server{rpc_handler}
  {}

  cryptonote::rpc::DaemonHandler rpc_handler;
  cryptonote::rpc::ZmqServer server;
};

struct t_internals {
private:
  t_protocol protocol;
public:
  t_core core;
  t_p2p p2p;
  std::vector<std::unique_ptr<t_rpc>> rpcs;
  std::unique_ptr<zmq_internals> zmq;

  t_internals(
      boost::program_options::variables_map const & vm
    )
    : core{vm}
    , protocol{vm, core, command_line::get_arg(vm, cryptonote::arg_offline)}
    , p2p{vm, protocol}
    , zmq{nullptr}
  {
    // Handle circular dependencies
    protocol.set_p2p_endpoint(p2p.get());
    core.set_protocol(protocol.get());

    const auto restricted = command_line::get_arg(vm, cryptonote::core_rpc_server::arg_restricted_rpc);
    const auto main_rpc_port = command_line::get_arg(vm, cryptonote::core_rpc_server::arg_rpc_bind_port);
    const auto restricted_rpc_port_arg = cryptonote::core_rpc_server::arg_rpc_restricted_bind_port;
    const bool has_restricted_rpc_port_arg = !command_line::is_arg_defaulted(vm, restricted_rpc_port_arg);
    rpcs.emplace_back(new t_rpc{vm, core, p2p, restricted, main_rpc_port, "core", !has_restricted_rpc_port_arg});

    if(has_restricted_rpc_port_arg)
    {
      auto restricted_rpc_port = command_line::get_arg(vm, restricted_rpc_port_arg);
      rpcs.emplace_back(new t_rpc{vm, core, p2p, true, restricted_rpc_port, "restricted", true});
    }

    if (!command_line::get_arg(vm, daemon_args::arg_zmq_rpc_disabled))
    {
      zmq.reset(new zmq_internals{core, p2p});

      const std::string zmq_port = command_line::get_arg(vm, daemon_args::arg_zmq_rpc_bind_port);
      const std::string zmq_address = command_line::get_arg(vm, daemon_args::arg_zmq_rpc_bind_ip);

      if (!zmq->server.init_rpc(zmq_address, zmq_port))
        throw std::runtime_error{"Failed to add TCP socket(" + zmq_address + ":" + zmq_port + ") to ZMQ RPC Server"};

      std::shared_ptr<cryptonote::listener::zmq_pub> shared;
      const std::vector<std::string> zmq_pub = command_line::get_arg(vm, daemon_args::arg_zmq_pub);
      if (!zmq_pub.empty() && !(shared = zmq->server.init_pub(epee::to_span(zmq_pub))))
        throw std::runtime_error{"Failed to initialize zmq_pub"};

      if (shared)
      {
        core.get().get_blockchain_storage().add_block_notify(cryptonote::listener::zmq_pub::chain_main{shared});
        core.get().get_blockchain_storage().add_miner_notify(cryptonote::listener::zmq_pub::miner_data{shared});
        core.get().add_sync_notify(cryptonote::listener::zmq_pub::sync{shared});
        core.get().set_txpool_listener(cryptonote::listener::zmq_pub::txpool_add{shared});
      }
    }
    else // if --no-zmq specified
    {
      // Assert that none of --zmq-rpc-bind-port, --zmq-rpc-bind-ip, and --zmq-pub are specified b/c
      // that does not make semantic sense with --no-zmq.
      if (command_line::get_arg(vm, daemon_args::arg_zmq_rpc_bind_port) !=
          daemon_args::arg_zmq_rpc_bind_port.default_value)
      {
        MWARNING("WARN: --zmq-rpc-bind-port has no effect because --no-zmq was specified");
      }
      else if (command_line::get_arg(vm, daemon_args::arg_zmq_rpc_bind_ip) !=
          daemon_args::arg_zmq_rpc_bind_ip.default_value)
      {
        MWARNING("WARN: --zmq-rpc-bind-ip has no effect because --no-zmq was specified");
      }
      else if (!command_line::get_arg(vm, daemon_args::arg_zmq_pub).empty())
      {
        MWARNING("WARN: --zmq-pub has no effect because --no-zmq was specified");
      }
    }
  }
};

void t_daemon::init_options(boost::program_options::options_description & option_spec)
{
  t_core::init_options(option_spec);
  t_p2p::init_options(option_spec);
  t_rpc::init_options(option_spec);
}

t_daemon::t_daemon(
    boost::program_options::variables_map const & vm,
    uint16_t public_rpc_port
  )
  : mp_internals{new t_internals{vm}},
  public_rpc_port(public_rpc_port)
{
}

t_daemon::~t_daemon() = default;

// MSVC is brain-dead and can't default this...
t_daemon::t_daemon(t_daemon && other)
{
  if (this != &other)
  {
    mp_internals = std::move(other.mp_internals);
    other.mp_internals.reset(nullptr);
    public_rpc_port = other.public_rpc_port;
  }
}

// or this
t_daemon & t_daemon::operator=(t_daemon && other)
{
  if (this != &other)
  {
    mp_internals = std::move(other.mp_internals);
    other.mp_internals.reset(nullptr);
    public_rpc_port = other.public_rpc_port;
  }
  return *this;
}

bool t_daemon::run(bool interactive)
{
  if (nullptr == mp_internals)
  {
    throw std::runtime_error{"Can't run stopped daemon"};
  }

  std::atomic<bool> stop(false), shutdown(false);
  boost::thread stop_thread = boost::thread([&stop, &shutdown, this] {
    while (!stop)
      epee::misc_utils::sleep_no_w(100);
    if (shutdown)
      this->stop_p2p();
  });
  epee::misc_utils::auto_scope_leave_caller scope_exit_handler = epee::misc_utils::create_scope_leave_handler([&](){
    stop = true;
    stop_thread.join();
  });
  tools::signal_handler::install([&stop, &shutdown](int){ stop = shutdown = true; });

  try
  {
    if (!mp_internals->core.run())
      return false;

    for(auto& rpc: mp_internals->rpcs)
      rpc->run();

    std::unique_ptr<daemonize::t_command_server> rpc_commands;
    if (interactive && mp_internals->rpcs.size())
    {
      // The first three variables are not used when the fourth is false
      rpc_commands.reset(new daemonize::t_command_server(0, 0, boost::none, epee::net_utils::ssl_support_t::e_ssl_support_disabled, false, mp_internals->rpcs.front()->get_server()));
      rpc_commands->start_handling(std::bind(&daemonize::t_daemon::stop_p2p, this));
    }

    if (mp_internals->zmq)
      mp_internals->zmq->server.run();
    else
      MINFO("ZMQ server disabled");

    if (public_rpc_port > 0)
    {
      MGINFO("Public RPC port " << public_rpc_port << " will be advertised to other peers over P2P");
      mp_internals->p2p.get().set_rpc_port(public_rpc_port);
    }
    
    mp_internals->p2p.run(); // blocks until p2p goes down

    if (rpc_commands)
      rpc_commands->stop_handling();

    if (mp_internals->zmq)
      mp_internals->zmq->server.stop();

    for(auto& rpc : mp_internals->rpcs)
      rpc->stop();
    MGINFO("Node stopped.");
    return true;
  }
  catch (std::exception const & ex)
  {
    MFATAL("Uncaught exception! " << ex.what());
    return false;
  }
  catch (...)
  {
    MFATAL("Uncaught exception!");
    return false;
  }
}

void t_daemon::stop()
{
  if (nullptr == mp_internals)
  {
    throw std::runtime_error{"Can't stop stopped daemon"};
  }
  mp_internals->p2p.stop();
  for(auto& rpc : mp_internals->rpcs)
    rpc->stop();

  mp_internals.reset(nullptr); // Ensure resources are cleaned up before we return
}

void t_daemon::stop_p2p()
{
  if (nullptr == mp_internals)
  {
    throw std::runtime_error{"Can't send stop signal to a stopped daemon"};
  }
  mp_internals->p2p.get().send_stop_signal();
}

} // namespace daemonize

File: utils/python-rpc/framework/daemon.py
# Copyright (c) 2018-2022, The Monero Project

# 
# All rights reserved.
# 
# Redistribution and use in source and binary forms, with or without modification, are
# permitted provided that the following conditions are met:
# 
# 1. Redistributions of source code must retain the above copyright notice, this list of
#    conditions and the following disclaimer.
# 
# 2. Redistributions in binary form must reproduce the above copyright notice, this list
#    of conditions and the following disclaimer in the documentation and/or other
#    materials provided with the distribution.
# 
# 3. Neither the name of the copyright holder nor the names of its contributors may be
#    used to endorse or promote products derived from this software without specific
#    prior written permission.
# 
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
# THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
# STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
# THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

"""Daemon class to make rpc calls and store state."""

from .rpc import JSONRPC 

class Daemon(object):

    def __init__(self, protocol='http', host='127.0.0.1', port=0, idx=0, restricted_rpc = False):
        base = 18480 if restricted_rpc else 18180
        self.host = host
        self.port = port
        self.rpc = JSONRPC('{protocol}://{host}:{port}'.format(protocol=protocol, host=host, port=port if port else base+idx))

    def getblocktemplate(self, address, prev_block = "", client = ""):
        getblocktemplate = {
            'method': 'getblocktemplate',
            'params': {
                'client': client,
                'wallet_address': address,
                'reserve_size' : 1,
                'prev_block' : prev_block,
            },
            'jsonrpc': '2.0', 
            'id': '0'
        }
        return self.rpc.send_json_rpc_request(getblocktemplate)
    get_block_template = getblocktemplate

    def get_miner_data(self):
        get_miner_data = {
            'method': 'get_miner_data',
            'jsonrpc': '2.0', 
            'id': '0'
        }
        return self.rpc.send_json_rpc_request(get_miner_data)

    def calc_pow(self, major_version, height, block_blob, seed_hash = ''):
        calc_pow = {
            'method': 'calc_pow',
            'params': {
                'major_version': major_version,
                'height': height,
                'block_blob' : block_blob,
                'seed_hash' : seed_hash,
            },
            'jsonrpc': '2.0',
            'id': '0'
        }
        return self.rpc.send_json_rpc_request(calc_pow)

    def add_aux_pow(self, blocktemplate_blob, aux_pow, client = ""):
        add_aux_pow = {
            'method': 'add_aux_pow',
            'params': {
                'blocktemplate_blob': blocktemplate_blob,
                'aux_pow' : aux_pow,
                'client' : client,
            },
            'jsonrpc': '2.0',
            'id': '0'
        }
        return self.rpc.send_json_rpc_request(add_aux_pow)

    def send_raw_transaction(self, tx_as_hex, do_not_relay = False, do_sanity_checks = True, client = ""):
        send_raw_transaction = {
            'client': client,
            'tx_as_hex': tx_as_hex,
            'do_not_relay': do_not_relay,
            'do_sanity_checks': do_sanity_checks,
        }
        return self.rpc.send_request("/send_raw_transaction", send_raw_transaction)
    sendrawtransaction = send_raw_transaction

    def submitblock(self, block):
        submitblock = {
            'method': 'submitblock',
            'params': [ block ],    
            'jsonrpc': '2.0', 
            'id': '0'
        }    
        return self.rpc.send_json_rpc_request(submitblock)
    submit_block = submitblock

    def getblock(self, hash = '', height = 0, fill_pow_hash = False, client = ""):
        getblock = {
            'method': 'getblock',
            'params': {
                'client': client,
                'hash': hash,
                'height': height,
                'fill_pow_hash': fill_pow_hash,
            },
            'jsonrpc': '2.0', 
            'id': '0'
        }
        return self.rpc.send_json_rpc_request(getblock)
    get_block = getblock

    def getlastblockheader(self, client = ""):
        getlastblockheader = {
            'method': 'getlastblockheader',
            'params': {
                'client': client,
            },
            'jsonrpc': '2.0', 
            'id': '0'
        }
        return self.rpc.send_json_rpc_request(getlastblockheader)
    get_last_block_header = getlastblockheader

    def getblockheaderbyhash(self, hash = "", hashes = [], client = ""):
        getblockheaderbyhash = {
            'method': 'getblockheaderbyhash',
            'params': {
                'client': client,
                'hash': hash,
                'hashes': hashes,
            },
            'jsonrpc': '2.0', 
            'id': '0'
        }
        return self.rpc.send_json_rpc_request(getblockheaderbyhash)
    get_block_header_by_hash = getblockheaderbyhash

    def getblockheaderbyheight(self, height, client = ""):
        getblockheaderbyheight = {
            'method': 'getblockheaderbyheight',
            'params': {
                'client': client,
                'height': height,
            },
            'jsonrpc': '2.0', 
            'id': '0'
        }
        return self.rpc.send_json_rpc_request(getblockheaderbyheight)
    get_block_header_by_height = getblockheaderbyheight

    def getblockheadersrange(self, start_height, end_height, fill_pow_hash = False, client = ""):
        getblockheadersrange = {
            'method': 'getblockheadersrange',
            'params': {
                'client': client,
                'start_height': start_height,
                'end_height': end_height,
                'fill_pow_hash': fill_pow_hash,
            },
            'jsonrpc': '2.0', 
            'id': '0'
        }
        return self.rpc.send_json_rpc_request(getblockheadersrange)
    get_block_headers_range = getblockheadersrange

    def get_connections(self, client = ""):
        get_connections = {
            'client': client,
            'method': 'get_connections',
            'jsonrpc': '2.0', 
            'id': '0'
        }
        return self.rpc.send_json_rpc_request(get_connections)

    def get_info(self, client = ""):
        get_info = {
            'method': 'get_info',
            'params': {
                'client': client,
            },
            'jsonrpc': '2.0',
            'id': '0'
        }    
        return self.rpc.send_json_rpc_request(get_info)
    getinfo = get_info

    def hard_fork_info(self, client = ""):
        hard_fork_info = {
            'method': 'hard_fork_info',
            'params': {
                'client': client,
            },
            'jsonrpc': '2.0',
            'id': '0'
        }    
        return self.rpc.send_json_rpc_request(hard_fork_info)

    def generateblocks(self, address, blocks=1, prev_block = "", starting_nonce = 0):
        generateblocks = {
            'method': 'generateblocks',
            'params': {
                'amount_of_blocks' : blocks,
                'reserve_size' : 20,
                'wallet_address': address,
                'prev_block': prev_block,
                'starting_nonce': starting_nonce,
            },
            'jsonrpc': '2.0', 
            'id': '0'
        }
        return self.rpc.send_json_rpc_request(generateblocks)

    def get_height(self, client = ""):
        get_height = {
                'method': 'get_height',
                'jsonrpc': '2.0',
                'id': '0'
        }
        return self.rpc.send_request("/get_height", get_height)
    getheight = get_height

    def pop_blocks(self, nblocks = 1):
        pop_blocks = {
            'nblocks' : nblocks,
        }
        return self.rpc.send_request("/pop_blocks", pop_blocks)

    def start_mining(self, miner_address, threads_count = 0, do_background_mining = False, ignore_battery = False):
        start_mining = {
            'miner_address' : miner_address,
            'threads_count' : threads_count,
            'do_background_mining' : do_background_mining,
            'ignore_battery' : ignore_battery,
        }
        return self.rpc.send_request('/start_mining', start_mining)

    def stop_mining(self):
        stop_mining = {
        }
        return self.rpc.send_request('/stop_mining', stop_mining)

    def mining_status(self):
        mining_status = {
        }
        return self.rpc.send_request('/mining_status', mining_status)

    def get_transaction_pool(self, client = ""):
        get_transaction_pool = {
            'client': client,
        }
        return self.rpc.send_request('/get_transaction_pool', get_transaction_pool)

    def get_transaction_pool_hashes(self, client = ""):
        get_transaction_pool_hashes = {
            'client': client,
        }
        return self.rpc.send_request('/get_transaction_pool_hashes', get_transaction_pool_hashes)

    def get_transaction_pool_stats(self, client = ""):
        get_transaction_pool_stats = {
            'client': client,
        }
        return self.rpc.send_request('/get_transaction_pool_stats', get_transaction_pool_stats)

    def flush_txpool(self, txids = []):
        flush_txpool = {
            'method': 'flush_txpool',
            'params': {
                'txids': txids
            },
            'jsonrpc': '2.0',
            'id': '0'
        }
        return self.rpc.send_json_rpc_request(flush_txpool)

    def get_version(self):
        get_version = {
            'method': 'get_version',
            'jsonrpc': '2.0',
            'id': '0'
        }
        return self.rpc.send_json_rpc_request(get_version)

    def get_bans(self):
        get_bans = {
            'method': 'get_bans',
            'params': {
            },
            'jsonrpc': '2.0',
            'id': '0'
        }
        return self.rpc.send_json_rpc_request(get_bans)

    def set_bans(self, bans = []):
        set_bans = {
            'method': 'set_bans',
            'params': {
                'bans': bans
            },
            'jsonrpc': '2.0',
            'id': '0'
        }
        return self.rpc.send_json_rpc_request(set_bans)

    def banned(self, address = ''):
        banned = {
            'method': 'banned',
            'params': {
                'address': address
            },
            'jsonrpc': '2.0',
            'id': '0'
        }
        return self.rpc.send_json_rpc_request(banned)

    def set_bootstrap_daemon(self, address, username = '', password = ''):
        set_bootstrap_daemon = {
            'address': address,
            'username': username,
            'password': password,
        }
        return self.rpc.send_request('/set_bootstrap_daemon', set_bootstrap_daemon)

    def get_public_nodes(self, gray = False, white = True):
        get_public_nodes = {
            'gray': gray,
            'white': white,
        }
        return self.rpc.send_request('/get_public_nodes', get_public_nodes)

    def get_transactions(self, txs_hashes = [], decode_as_json = False, prune = False, split = False, client = ""):
        get_transactions = {
            'client': client,
            'txs_hashes': txs_hashes,
            'decode_as_json': decode_as_json,
            'prune': prune,
            'split': split,
        }
        return self.rpc.send_request('/get_transactions', get_transactions)
    gettransactions = get_transactions

    def get_outs(self, outputs = [], get_txid = False, client = ""):
        get_outs = {
            'client': client,
            'outputs': outputs,
            'get_txid': get_txid,
        }
        return self.rpc.send_request('/get_outs', get_outs)

    def get_coinbase_tx_sum(self, height, count, client = ""):
        get_coinbase_tx_sum = {
            'method': 'get_coinbase_tx_sum',
            'params': {
                'client': client,
                'height': height,
                'count': count,
            },
            'jsonrpc': '2.0',
            'id': '0'
        }
        return self.rpc.send_json_rpc_request(get_coinbase_tx_sum)

    def get_output_distribution(self, amounts = [], from_height = 0, to_height = 0, cumulative = False, binary = False, compress = False, client = ""):
        get_output_distribution = {
            'method': 'get_output_distribution',
            'params': {
                'client': client,
                'amounts': amounts,
                'from_height': from_height,
                'to_height': to_height,
                'cumulative': cumulative,
                'binary': binary,
                'compress': compress,
            },
            'jsonrpc': '2.0',
            'id': '0'
        }
        return self.rpc.send_json_rpc_request(get_output_distribution)

    def get_output_histogram(self, amounts = [], min_count = 0, max_count = 0, unlocked = False, recent_cutoff = 0, client = ""):
        get_output_histogram = {
            'method': 'get_output_histogram',
            'params': {
                'client': client,
                'amounts': amounts,
                'min_count': min_count,
                'max_count': max_count,
                'unlocked': unlocked,
                'recent_cutoff': recent_cutoff,
            },
            'jsonrpc': '2.0',
            'id': '0'
        }
        return self.rpc.send_json_rpc_request(get_output_histogram)

    def set_log_level(self, level):
        set_log_level = {
            'level': level,
        }
        return self.rpc.send_request('/set_log_level', set_log_level)

    def set_log_categories(self, categories = ''):
        set_log_categories = {
            'categories': categories,
        }
        return self.rpc.send_request('/set_log_categories', set_log_categories)

    def get_alt_blocks_hashes(self, client = ""):
        get_alt_blocks_hashes = {
            'client': client,
        }
        return self.rpc.send_request('/get_alt_blocks_hashes', get_alt_blocks_hashes)

    def get_alternate_chains(self, client = ""):
        get_alternate_chains = {
            'method': 'get_alternate_chains',
            'params': {
                'client': client,
            },
            'jsonrpc': '2.0',
            'id': '0'
        }
        return self.rpc.send_json_rpc_request(get_alternate_chains)

    def get_fee_estimate(self, grace_blocks = 0):
        get_fee_estimate = {
            'method': 'get_fee_estimate',
            'params': {
                'grace_blocks': grace_blocks,
            },
            'jsonrpc': '2.0',
            'id': '0'
        }
        return self.rpc.send_json_rpc_request(get_fee_estimate)

    def is_key_image_spent(self, key_images = [], client = ""):
        is_key_image_spent = {
            'key_images': key_images,
            'client': client,
        }
        return self.rpc.send_request('/is_key_image_spent', is_key_image_spent)

    def save_bc(self):
        save_bc = {
        }
        return self.rpc.send_request('/save_bc', save_bc)

    def get_peer_list(self):
        get_peer_list = {
        }
        return self.rpc.send_request('/get_peer_list', get_peer_list)

    def set_log_hash_rate(self, visible):
        set_log_hash_rate = {
            'visible': visible,
        }
        return self.rpc.send_request('/set_log_hash_rate', set_log_hash_rate)

    def stop_daemon(self):
        stop_daemon = {
        }
        return self.rpc.send_request('/stop_daemon', stop_daemon)

    def get_net_stats(self):
        get_net_stats = {
        }
        return self.rpc.send_request('/get_net_stats', get_net_stats)

    def get_limit(self):
        get_limit = {
        }
        return self.rpc.send_request('/get_limit', get_limit)

    def set_limit(self, limit_down, limit_up):
        set_limit = {
            'limit_down': limit_down,
            'limit_up': limit_up,
        }
        return self.rpc.send_request('/set_limit', set_limit)

    def out_peers(self, out_peers):
        out_peers = {
            'out_peers': out_peers,
        }
        return self.rpc.send_request('/out_peers', out_peers)

    def in_peers(self, in_peers):
        in_peers = {
            'in_peers': in_peers,
        }
        return self.rpc.send_request('/in_peers', in_peers)

    def update(self, command, path = None):
        update = {
            'command': command,
            'path': path,
        }
        return self.rpc.send_request('/update', update)

    def get_block_count(self):
        get_block_count = {
            'method': 'get_block_count',
            'params': {
            },
            'jsonrpc': '2.0', 
            'id': '0'
        }
        return self.rpc.send_json_rpc_request(get_block_count)
    getblockcount = get_block_count

    def get_block_hash(self, height):
        get_block_hash = {
            'method': 'get_block_hash',
            'params': [height],
            'jsonrpc': '2.0', 
            'id': '0'
        }
        return self.rpc.send_json_rpc_request(get_block_hash)
    on_get_block_hash = get_block_hash
    on_getblockhash = get_block_hash

    def relay_tx(self, txids = [], client = ""):
        relay_tx = {
            'method': 'relay_tx',
            'params': {
                'txids': txids,
                'client': client,
            },
            'jsonrpc': '2.0', 
            'id': '0'
        }
        return self.rpc.send_json_rpc_request(relay_tx)

    def sync_info(self, client = ""):
        sync_info = {
            'method': 'sync_info',
            'params': {
                'client': client,
            },
            'jsonrpc': '2.0', 
            'id': '0'
        }
        return self.rpc.send_json_rpc_request(sync_info)

    def get_txpool_backlog(self, client = ""):
        get_txpool_backlog = {
            'method': 'get_txpool_backlog',
            'params': {
                'client': client,
            },
            'jsonrpc': '2.0', 
            'id': '0'
        }
        return self.rpc.send_json_rpc_request(get_txpool_backlog)

    def prune_blockchain(self, check = False):
        prune_blockchain = {
            'method': 'prune_blockchain',
            'params': {
                'check': check,
            },
            'jsonrpc': '2.0', 
            'id': '0'
        }
        return self.rpc.send_json_rpc_request(prune_blockchain)

    def flush_cache(self, bad_txs = False):
        flush_cache = {
            'method': 'flush_cache',
            'params': {
                'bad_txs': bad_txs,
            },
            'jsonrpc': '2.0',
            'id': '0'
        }
        return self.rpc.send_json_rpc_request(flush_cache)

    def sync_txpool(self):
        sync_txpool = {
            'method': 'sync_txpool',
            'params': {
            },
            'jsonrpc': '2.0',
            'id': '0'
        }
        return self.rpc.send_json_rpc_request(sync_txpool)

    def rpc_access_info(self, client):
        rpc_access_info = {
            'method': 'rpc_access_info',
            'params': {
                'client': client,
            },
            'jsonrpc': '2.0',
            'id': '0'
        }
        return self.rpc.send_json_rpc_request(rpc_access_info)

    def rpc_access_submit_nonce(self, client, nonce, cookie):
        rpc_access_submit_nonce = {
            'method': 'rpc_access_submit_nonce',
            'params': {
                'client': client,
                'nonce': nonce,
                'cookie': cookie,
            },
            'jsonrpc': '2.0',
            'id': '0'
        }
        return self.rpc.send_json_rpc_request(rpc_access_submit_nonce)

    def rpc_access_pay(self, client, paying_for, payment):
        rpc_access_pay = {
            'method': 'rpc_access_pay',
            'params': {
                'client': client,
                'paying_for': paying_for,
                'payment': payment,
            },
            'jsonrpc': '2.0',
            'id': '0'
        }
        return self.rpc.send_json_rpc_request(rpc_access_pay)

    def rpc_access_tracking(self, clear = False):
        rpc_access_tracking = {
            'method': 'rpc_access_tracking',
            'params': {
                'clear': clear,
            },
            'jsonrpc': '2.0',
            'id': '0'
        }
        return self.rpc.send_json_rpc_request(rpc_access_tracking)

    def rpc_access_data(self):
        rpc_access_data = {
            'method': 'rpc_access_data',
            'params': {
            },
            'jsonrpc': '2.0',
            'id': '0'
        }
        return self.rpc.send_json_rpc_request(rpc_access_data)

    def rpc_access_account(self, client, delta_balance = 0):
        rpc_access_account = {
            'method': 'rpc_access_account',
            'params': {
                'client': client,
                'delta_balance': delta_balance,
            },
            'jsonrpc': '2.0',
            'id': '0'
        }
        return self.rpc.send_json_rpc_request(rpc_access_account)
